### PR TITLE
Bdc/feature/improve data handling

### PR DESCRIPTION
**Pull request by luccalb**
- Merged pipeline_dal.py into pipeline.py
- Renamed the DAL components
- Save snapshots of the dataframe to S3/local after each step

**Future ideas:**

- Delete the snapshots automatically if the pipeline completely runs without errors (stub `Repository.clean_snapshots()` already exists)
- At the beginning of a pipeline run, check if the SNAPSHOTS location is not empty and ask the user if he wants to run the pipeline on the latest snapshot
- Snapshot the dataframe after processing n leads inside of a step (could make sense if we process up to 100k leads per step)
- Remember for every lead what steps were run and which weren't (e.g. by adding a column`{step_name}_executed={True|False|None}`

### CODE REVIEW

**Comment by soapyheas** [comment]
Consulted with Lucca before adding the following:
* Added a try/except block when using boto3.client.copy to catch errors related to trying to copy a file that does not exist yet.

**Review by rbbozkurt — APPROVED** [review]
Looks pretty fine, well done!

File: src/demos.py [hunk 1]
```diff
@@ -186,12 +186,13 @@ def pipeline_demo():
...
-        choice = input(
-            "The output cannot be limited when uploading to S3.\nThe limit will be removed, and the pipeline will be executed on the full database.\n\nWould you like to continue? (y/n)\n"
-        )
-
-        if choice != "y" and choice != "Y":
+    if (
+        limit is not None
+        and get_database().DF_OUTPUT == "s3://amos--data--events/leads/enriched.csv"
+    ):
+        if get_yes_no_input(
```
Comment (rbbozkurt) [inline src/demos.py:193]: We should not be proceeding with the execution if answer is "no". 
```suggestion
        if get_yes_no_input(
            f"The output cannot be limited when uploading to {get_database().DF_OUTPUT}.\nThe limit will be removed, and the pipeline will be executed on the full database.\n\nWould you like to continue? (y/n)\n"
        ):
            limit = None
        else:
            return
```

**Review by rbbozkurt — CHANGES_REQUESTED** [review]

File: Pipfile [hunk 2]
```diff
@@ -10,6 +10,7 @@ name = "pypi"
 pytest = "==7.4.0"
 pre-commit = "==3.5.0"
 flake8 = "==6.0.0"
+pytest-env = "*"
```
[the remark this answers is not in the document]
Comment (luccalb) [reply Pipfile]: fixed

File: src/demos.py (hunk 1 above)
Comment (luccalb) [reply src/demos.py:193]: fixed

**Pull request merged**